subsys/bluetooth/audio/has.c

```diff
@@ -19,13 +19,45 @@
 #define LOG_MODULE_NAME bt_has
 #include "common/log.h"
 
+/* HAS requires encryption key with at least 128 bits of entropy */
+#define HAS_ENC_KEY_SIZE_MIN 16
+
 static struct bt_has has;
 
+static int check_security(struct bt_conn *conn)
+{
+	/* The security requirements for all characteristics is LE Security Mode 1 Level 2 */
+	BT_ASSERT(bt_conn_get_security(conn) > BT_SECURITY_L1);
+
+	if (bt_conn_enc_key_size(conn) < HAS_ENC_KEY_SIZE_MIN) {
+		return BT_GATT_ERR(BT_ATT_ERR_ENCRYPTION_KEY_SIZE);
+	}
+
+	/* Access to all characteristics (...) require an encryption key derived from any of the
+	 * following:
+	 * • LE Secure Connections
+	 * • BR/EDR Secure Connections (if Cross-Transport Key Derivation is used)
+	 * • OOB method
+	 *
+	 * TODO: Check the encryption method.
+	 *       What error shall be returned if requirement above is not met??
+	 */
+
+	return 0;
+}
+
 static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
 			     uint16_t len, uint16_t offset)
 {
+	int err;
+
 	BT_DBG("conn %p attr %p offset %d", (void *)conn, attr, offset);
 
+	err = check_security(conn);
```

Comment:
Would it make more sense to add these additional security requirements to the GATT API, and have GATT handle it, like it currently handles the `BT_GATT_PERM_READ_` values?

Reply:
Is the only additional requirement a check for the encryption key size? I don't mind doing this on a GATT level (which seems better than what's right now proposed in this PR), but would it perhaps be even simpler to do this in Kconfig by adding something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`?

Reply:
> Is the only additional requirement a check for the encryption key size? I don't mind doing this on a GATT level (which seems better than what's right now proposed in this PR), but would it perhaps be even simpler to do this in Kconfig by adding something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`?

From the PR description: 
> Access to all characteristics (...) shall require an encryption
> key with at least 128 bits of entropy, derived from any of the following:
> • LE Secure Connections
> • BR/EDR Secure Connections (if Cross-Transport Key Derivationd)
> • OOB method

So are 2 requirements: 128-bit key, and then one of the 3 methods used. 

We've used something like `depends on BT_SMP_MIN_ENC_KEY_SIZE == 16`  in the past in the form of `select BT_SMP_SC_PAIR_ONLY` (or `depends on BT_SMP_SC_PAIR_ONLY`) to satisfy the LE secure connection requirement. 

However, this has the downside of being too strict: Technically the requirement is only the service characteristic access, and using `BT_SMP_SC_PAIR_ONLY` or `BT_SMP_MIN_ENC_KEY_SIZE` is a device-wide requirement. In most cases it should be fine, but we are adding additional requirements for non-"LE Audio" services that way

For the key size, we could add an additional value for the GATT characeristics that contain the minimum required key length. For the other requirements (LE secure connection, BR/EDR secure connections or OOB) it is going to be harder.

Reply:
@Thalley is the method stuff really that complicated? The only remaining method is legacy LE pairing, so all you really need to be able to exclude is that (then anything else satisfies the requirement), right?

Reply:
... in that sense it's probably sufficient to add a boolean option (or flag) to the characteristic definition

Reply:
Actually, we already have that `BT_SMP_SC_PAIR_ONLY` (and I see you already referenced it). I don't think there's a problem with that being a device-wide requirement, since (unlike `BT_SMP_SC_ONLY`) you can still have characteristics which don't require encryption, mixed which characteristics which do. It's probably also better from a user experience perspective, since you don't get weird situations where you've been able to pair, but only part of the services work - I think it's better to just fail the pairing in that case.

Reply:
The same argument can also be used for the encryption key size, i.e. that it's less confusing for the user to fail pairing rather than have a mix of services that work and those that don't work. So in that sense this should be solvable with the existing Kconfig options.

Reply:
I think that was also the conclusion last time we visited this issue, which is why we ended up doing `select BT_SMP_SC_PAIR_ONLY`. 

At the very least, it's is better to be more restrictive than not complying with security requirements :) 

`BT_SMP_SC_PAIR_ONLY` does not cover the OOB method though, and as mentioned I don't think it's worth dealing with the BR/EDR method. 

Personally I am fine with rejecting pairing that doesn't conform with the highest security requirement, even if it perhaps isn't considered the "GATT way". 

Any device that supports the LE Audio services will be advertising that, and in the case you are connecting to those devices, you should definitely be able to support LE Secure Connections and 128-bits anyhow.

Reply:
So the conclusion is to enable `BT_SMP_SC_PAIR_ONLY` option.
The PR has been updated.